### ISSUE TITLE
IllegalStateException: Already resumed, but proposed with update kotlin.Unit

### ISSUE THREAD

**Issue by oguzkocer** [wordpress-mobile/WordPress-Android#11620]
**Sentry Url**: https://sentry.io/share/issue/a1811baed63f46bc80b27ec6e7760fca/
**User Count**: 1
**Count**: 1
**First Release**: org.wordpress.android-14.6-rc-1
**First Seen**: 2020-04-08T11:16:30Z
**Last Release**: org.wordpress.android-14.6-rc-1
**Last Seen**: 2020-04-08T11:16:30Z
**24 Hours**: 1
**30 Days**: 1

**Stacktrace**:
```
IllegalStateException: Already resumed, but proposed with update kotlin.Unit
    at kotlinx.coroutines.CancellableContinuationImpl.alreadyResumedError(CancellableContinuationImpl.kt:332)
    at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:327)
    at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:247)
    at org.wordpress.android.viewmodel.pages.PagesViewModel.handlePageUpdated(PagesViewModel.kt:774)
    at org.wordpress.android.viewmodel.pages.PagesViewModel$start$1.invoke(PagesViewModel.kt:212)
    at org.wordpress.android.viewmodel.pages.PagesViewModel$start$1.invoke(PagesViewModel.kt:84)
    at org.wordpress.android.viewmodel.pages.PageListEventListener$onPostChanged$1.invokeSuspend(PageListEventListener.kt:102)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:561)
...
(3 additional frame(s) were not displayed)
```

**Comment by malinajirka** [comment]
@jd-alexander Could you please look into this. Thanks!

**Comment by jd-alexander** [comment]
@malinajirka I investigated this but I wasn't able to reproduce it. 

The behavior is supposed to be triggered when a post is uploaded so I got to that part by creating a post offline and then going online and the `postUploadStarted` event starts the continuation that we then resume. 

I realized that `postUploadStarted` is called twice for a single upload. Is that by design?

I also tried running `handlePageUpdated` multiple times to see if that would reproduce the issue but it didn't. 

I am wondering if this is simply happening because `handlePageUpdated` is being called multiple times and the continuation is not being removed quickly enough. I am wondering if synchronization is something that could fix this issue. 

I experimented with `continuation.context.isActive` to see if there's a way I could check to see if the continuation is active before calling resume but it doesn't seem to operate as expected. (Needs more investigation)

I was also wondering if we would need our own flag that we mark once we are within the `handlePageUpdated` function so `resume` isn't called twice 🤔 

These are all points up for discussion. Let me know what you think and what your investigation leads to.

**Comment by malinajirka** [comment]
Thanks for the investigation @jd-alexander!

I wasn't able to reproduce this issue either. However, I found a few issues with the implementation. 

tl;dr; Most of this logic has been there for ages and we just didn't realize we could remove it with the introduction of the overlay + UploadService on the page list. Wdyt?

1. We don't remove the item from the map on timeout - [here](https://github.com/wordpress-mobile/WordPress-Android/blob/2b2ea3eaa54b867914436880eda2266253682b19/WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt#L276). I think we'd need to add the following `?: pageUpdateContinuations.remove(id)`.
2. We use[ id = 0 for local posts](https://github.com/wordpress-mobile/WordPress-Android/blob/2b2ea3eaa54b867914436880eda2266253682b19/WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt#L768) as they don't have a remoteId. However, when there are multiple local posts all of them use the same id (0) so the code can't work.
3. We [try to resume the continuation whenever a post is updated](https://github.com/wordpress-mobile/WordPress-Android/blob/70cfd230bf970905196f756f57aa28f868fb6aed/WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt#L102). I think we should resume it only when an upload completes as PostUpdated is called way too often - eg every 500ms when you are editing a post in the editor.
4. All this logic isn't imo needed anymore. We don't need to suppress the more-buttons on the items for upload. We support multiple-concurrent uploads and the item which is being uploaded has disabled overlay, so the user can't click on the more-buttons on that particular item.
